geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/RangeHistogram.scala

```diff
@@ -0,0 +1,238 @@
+/***********************************************************************
+  * Copyright (c) 2013-2015 Commonwealth Computer Research, Inc.
+  * All rights reserved. This program and the accompanying materials
+  * are made available under the terms of the Apache License, Version 2.0 which
+  * accompanies this distribution and is available at
+  * http://www.opensource.org/licenses/apache2.0.php.
+  *************************************************************************/
+
+package org.locationtech.geomesa.utils.stats
+
+import java.lang
+import java.util.Date
+
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.util.parsing.json.JSONObject
+
+/**
+ * This object provides the type classes used by the RangeHistogram.
+ * BinAble provides methods which help bin a type's range of values into
+ * equal-sized (or close to equal-sized in many cases) smaller ranges.
+ */
+object BinHelper {
+  trait BinAble[T] {
+    def getBinSize(numBins: Int, lowerEndpoint: T, upperEndpoint: T): T
+    def getBinIndex(attributeValue: T, binSize: T, numBins: Int, lowerEndpoint: T, upperEndpoint: T): T
+    def getBinKey(binSize: T, bucketIndex: Int, lowerEndpoint: T): T
+  }
+
+  implicit object BinAbleDate extends BinAble[Date] {
+    override def getBinSize(numBins: Int, lowerEndpoint: Date, upperEndpoint: Date): Date = {
+      new Date((upperEndpoint.getTime - lowerEndpoint.getTime) / numBins)
+    }
+
+    override def getBinIndex(attributeValue: Date,
+                             binSize: Date,
+                             numBins: Int,
+                             lowerEndpoint: Date,
+                             upperEndpoint: Date): Date = {
+      if (attributeValue.getTime >= lowerEndpoint.getTime && attributeValue.getTime < upperEndpoint.getTime) {
```

Comment:
[style] dates have a .before and .after that would be a little cleaner than comparing times.